File: Client/iOS/ScreenShareExtension/SampleHandler.m
// Copyright (c) 2023 BytePlus Pte. Ltd.
// SPDX-License-Identifier: Apache-2.0
#import "SampleHandler.h"
#import <Appconfig/BuildConfig.h>
#import <VeLiveReplayKitExtension/VeLiveReplayKitExtension.h>
#import <BytePlusRTCScreenCapturer/ByteRTCScreenCapturerExt.h>

@interface SampleHandler()<VeLiveReplayKitExtensionDelegate, ByteRtcScreenCapturerExtDelegate>

@property (nonatomic, copy) NSString* stopMessage;

@property (nonatomic, copy) NSString* shareType;

@end

@implementation SampleHandler

- (instancetype)init {
    self = [super init];
    if (self) {
        _stopMessage = @"stopShare";
        _shareType = @"";
    }
    return self;
}

- (void)broadcastStartedWithSetupInfo:(NSDictionary<NSString *,NSObject *> *)setupInfo {
    NSUserDefaults  *screenShareUserDefaults = [[NSUserDefaults alloc] initWithSuiteName:APP_GROUP_ID];
    _shareType = [screenShareUserDefaults stringForKey:@"shareType"];
    // User has requested to start the broadcast. Setup info from the UI extension can be supplied but optional.
    if (self.shareType && [self.shareType isEqualToString:@"rtc"]) {
        [[ByteRtcScreenCapturerExt shared] startWithDelegate:self groupId:APP_GROUP_ID];
        [screenShareUserDefaults setValue:@"" forKey:@"shareType"];
        [screenShareUserDefaults synchronize];
    } else {
        [[VeLiveReplayKitExtension sharedInstance] startWithAppGroup:APP_GROUP_ID delegate:self];
    }
}

- (void)broadcastPaused {
    if (self.shareType && [self.shareType isEqualToString:@"rtc"]) {
        return;
    }
    [[VeLiveReplayKitExtension sharedInstance] broadcastPaused];
    // User has requested to pause the broadcast. Samples will stop being delivered.
}

- (void)broadcastResumed {
    if (self.shareType && [self.shareType isEqualToString:@"rtc"]) {
        return;
    }
    [[VeLiveReplayKitExtension sharedInstance] broadcastResumed];
    // User has requested to resume the broadcast. Samples delivery will resume.
}

- (void)broadcastFinished {
    // User has requested to finish the broadcast.
    if (self.shareType && [self.shareType isEqualToString:@"rtc"]) {
        [[ByteRtcScreenCapturerExt shared] stop];
    } else {
        [[VeLiveReplayKitExtension sharedInstance] broadcastFinished];
    }
}

- (void)processSampleBuffer:(CMSampleBufferRef)sampleBuffer withType:(RPSampleBufferType)sampleBufferType {
    if (self.shareType && [self.shareType isEqualToString:@"rtc"]) {
        [[ByteRtcScreenCapturerExt shared] processSampleBuffer:sampleBuffer withType:sampleBufferType];
    } else {
        [[VeLiveReplayKitExtension sharedInstance] processSampleBuffer:sampleBuffer withType:sampleBufferType];
    }
}


- (void)broadcastFinished:(VeLiveReplayKitExtension *)broadcast reason:(VeLiveReplayKitExtensionReason)reason {
    if (self.shareType && [self.shareType isEqualToString:@"rtc"]) {
        return;
    }
    NSString *tip = @"";
    switch (reason) {
        case VeLiveReplayKitExtensionReasonMainStop:
            tip = @"main app stop screen capture";
            break;
    }

    NSError *error = [NSError errorWithDomain:NSStringFromClass(self.class)
                                             code:0
                                         userInfo:@{
                                             NSLocalizedFailureReasonErrorKey:tip
                                         }];
    [self finishBroadcastWithError:error];

}

#pragma mark- ByteRtcScreenCapturerExtDelegate

- (void)onNotifyAppRunning {
    
}

- (void)onQuitFromApp {
    [self finishBroadcastWithError:[NSError errorWithDomain:RPRecordingErrorDomain
                                                       code:RPRecordingErrorUserDeclined
                                                   userInfo:@{
        NSLocalizedFailureReasonErrorKey : self.stopMessage,
    }]];
}

- (void)onReceiveMessageFromApp:(nonnull NSData *)message {
    self.stopMessage = [[NSString alloc] initWithData:message encoding:NSUTF8StringEncoding];
}

- (void)onSocketConnect {
    
}

- (void)onSocketDisconnect {
    [[ByteRtcScreenCapturerExt shared] stop];
    [self onQuitFromApp];
}

@end

File: RTCAPIExmapleDemo/iOS/ApiExample/ApiExampleScreenShare/SampleHandler.swift
import ReplayKit
import Foundation
import BytePlusRTC

class SampleHandler: RPBroadcastSampleHandler, ByteRtcScreenCapturerExtDelegate {
    

    override func broadcastStarted(withSetupInfo setupInfo: [String : NSObject]?) {
        // User has requested to start the broadcast. Setup info from the UI extension can be supplied but optional.
        print("-------")
        ByteRtcScreenCapturerExt.shared.start(with: self, groupId: APP_GROUP);
        
    }
    
    override func broadcastPaused() {
        // User has requested to pause the broadcast. Samples will stop being delivered.
    }
    
    override func broadcastResumed() {
        // User has requested to resume the broadcast. Samples delivery will resume.
    }
    
    override func broadcastFinished() {
        ByteRtcScreenCapturerExt.shared.stop()
    }
    
    override func processSampleBuffer(_ sampleBuffer: CMSampleBuffer, with sampleBufferType: RPSampleBufferType) {
        switch sampleBufferType {
            
        case .video, .audioApp: // 采集到的屏幕视频流 和 屏幕音频流
            ByteRtcScreenCapturerExt.shared.processSampleBuffer(sampleBuffer, with: sampleBufferType)
            
            break
        case .audioMic:
            // Handle audio sample buffer for mic audio
            break
        @unknown default:
            // Handle other sample buffer types
            fatalError("Unknown type of sample buffer")
        }
    }
    
    // MARK: ByteRtcScreenCapturerExtDelegate
    func onQuitFromApp() {
        let userInfo = [
            NSLocalizedFailureReasonErrorKey: NSLocalizedString("stopScreenSharing", comment:"")
        ]
        let error = NSError(domain: RPRecordingErrorDomain,
                            code: RPRecordingErrorCode.userDeclined.rawValue,
                            userInfo: userInfo)
        self.finishBroadcastWithError(error)

    }
    
    func onReceiveMessage(fromApp message: Data) {
        print("SampleHandler onReceiveMessageFromApp")
    }
    
    func onSocketDisconnect() {
        print("SampleHandler onSocketDisconnect")
    }
    
    func onSocketConnect() {
        print("SampleHandler onSocketConnect")
    }
    
    func onNotifyAppRunning() {
        print("SampleHandler onNotifyAppRunning")
    }
}
